src/basic/recurse-dir.c

```diff
@@ -104,6 +105,14 @@ int readdir_all(int dir_fd,
                 if (ignore_dirent(entry, flags))
                         continue;
 
+                if (FLAGS_SET(flags, RECURSE_DIR_ENSURE_TYPE)) {
+                        r = dirent_ensure_type(dir_fd, entry);
+                        if (r == -ENOENT)
+                                continue;
```

Comment:
de->n_entries won't match reality now, which will blow up in the typesafe_qsort() case below.

in fact, maybe we should add an `assert(j == de->n_entries)` after this loop to ensure everything is properly initialized as we presume it is.